### PR TITLE
[feature-wip] (memory tracker) (step3) Switch TLS mem tracker to separate more detailed memory usage

### LINKED ISSUE #7196: [Feature] Refactored memory statistics framework MemTracker
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

https://shimo.im/docs/DT6JXDRkdTvdyV3G

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by xinyiZzz**
# Proposed changes

Issue Number: close #7196 (step 3.1/3)

## Problem Summary:

In pr #8476, all memory usage of a process is recorded in the process mem tracker, and all memory usage of a query is recorded in the query mem tracker, and it is still necessary to manually call `transfer to` to track the cached memory size.

We hope to separate out more detailed memory usage based on Hook TCMalloc new/delete + TLS mem tracker.

In this pr, the more detailed mem tracker is switched to TLS, which automatically and accurately counts more detailed memory usage than before.
## Checklist(Required)

1. Does it affect the original behavior: (Yes)
2. Has unit tests been added: (Yes)
3. Has document been added or modified: (No)
4. Does it need to update dependencies: (No)
5. Are there any changes that cannot be rolled back: (No)

## Further comments

1. Added the method of switching mem tracker to TLS;
- More accurate statistics of cache memory size. It is no longer necessary to manually calculate the memory size of each add/free cache, which can easily miss memory such as vector resize.
- (TODO) Separate out the memory usage of each operator from the query mem tracker.
2. Added callback behavior when switching TLS mem tracker consume fails;
- Replace all methods of manually determining whether the instance mem tracker has exceeded limit in the loop, such as join node and agg node.

Statistical effect:
- Initially (after BE restart)
image

- Query is executing
image
- After Query is over
image

- Load is executing
Note: LoadChannelMgr consumption is 0, because this pr has not really called switch tls mem tracker, and the next pr will be modified in batches.
image

- After Load is over
image

Note: If QueryPool and LoadPool are found to have negative values for a short time, this is normal, refer to the comments in `MemTrackerTaskPool::logout_task_mem_tracker`.

### CODE REVIEW

File: be/src/exec/cross_join_node.cpp [hunk 1]
```diff
@@ -52,6 +53,7 @@ Status CrossJoinNode::close(RuntimeState* state) {
 Status CrossJoinNode::construct_build_side(RuntimeState* state) {
     // Do a full scan of child(1) and store all build row batches.
     RETURN_IF_ERROR(child(1)->open(state));
+    SCOPED_SWITCH_THREAD_LOCAL_MEM_TRACKER_ERR_CB("Cross join, while getting next from the child 1");
```
Comment (yangzhg) [inline be/src/exec/cross_join_node.cpp:56]:
```suggestion
    SCOPED_SWITCH_THREAD_LOCAL_MEM_TRACKER_ERR_CB("Cross join, while getting next from child 1");
```

File: be/src/runtime/thread_mem_tracker_mgr.h [hunk 2]
```diff
@@ -80,7 +72,7 @@ class ThreadMemTrackerMgr {
     }
 
     void clear_untracked_mems() {
-        for(auto untracked_mem : _untracked_mems) {
+        for (auto untracked_mem : _untracked_mems) {
```
Comment (yangzhg) [inline be/src/runtime/thread_mem_tracker_mgr.h:75]:
```suggestion
        for (const auto& untracked_mem : _untracked_mems) {
```

File: be/src/runtime/thread_mem_tracker_mgr.h [hunk 3]
```diff
@@ -101,13 +93,36 @@ class ThreadMemTrackerMgr {
...
+    }
+
     // Note that, If call the memory allocation operation in TCMalloc new/delete Hook,
     // such as calling LOG/iostream/sstream/stringstream/etc. related methods,
     // must increase the control to avoid entering infinite recursion, otherwise it may cause crash or stuck,
     void cache_consume(int64_t size);
 
     void noncache_consume();
 
+    bool attach_task() { return _task_id != ""; }
```
Comment (yangzhg) [inline be/src/runtime/thread_mem_tracker_mgr.h:124]: Rename this function

File: be/src/vec/exec/vcross_join_node.cpp [hunk 4]
```diff
@@ -53,6 +54,7 @@ Status VCrossJoinNode::close(RuntimeState* state) {
 Status VCrossJoinNode::construct_build_side(RuntimeState* state) {
     // Do a full scan of child(1) and store all build row batches.
     RETURN_IF_ERROR(child(1)->open(state));
+    SCOPED_SWITCH_THREAD_LOCAL_MEM_TRACKER_ERR_CB("Cross join, while getting next from the child 1");
```
Comment (yangzhg) [inline be/src/vec/exec/vcross_join_node.cpp:57]:
```suggestion
    SCOPED_SWITCH_THREAD_LOCAL_MEM_TRACKER_ERR_CB("Vec Cross join, while getting next from child 1");
```

File: be/src/vec/exec/vset_operation_node.cpp [hunk 5]
```diff
@@ -228,6 +229,8 @@ void VSetOperationNode::hash_table_init() {
 //build a hash table from child(0)
 Status VSetOperationNode::hash_table_build(RuntimeState* state) {
     RETURN_IF_ERROR(child(0)->open(state));
+    SCOPED_SWITCH_THREAD_LOCAL_MEM_TRACKER_ERR_CB(
+                "Set Operation Node, while constructing the hash table");
```
Comment (yangzhg) [inline be/src/vec/exec/vset_operation_node.cpp:233]:
```suggestion
                "Vec Set Operation Node, while constructing the hash table");
```

File: be/src/runtime/thread_mem_tracker_mgr.cpp [hunk 6]
```diff
@@ -70,12 +72,12 @@ void ThreadMemTrackerMgr::exceeded_cancel_task(const std::string& cancel_details
 
 void ThreadMemTrackerMgr::exceeded(int64_t mem_usage, Status st) {
     auto rst = _mem_trackers[_tracker_id]->mem_limit_exceeded(
-            nullptr, "In TCMalloc Hook, " + _consume_err_call_back.action_type, mem_usage, st);
-    if (_consume_err_call_back.call_back_func != nullptr) {
-        _consume_err_call_back.call_back_func();
+            nullptr, "In TCMalloc Hook, " + _consume_err_cb.cancel_msg, mem_usage, st);
```
Comment (yangzhg) [inline be/src/runtime/thread_mem_tracker_mgr.cpp:75]:
```suggestion
            nullptr, fmt::format("In TCMalloc Hook, {}", _consume_err_cb.cancel_msg), mem_usage, st);
```

File: be/src/exec/cross_join_node.cpp (hunk 1 above)
Comment (xinyiZzz) [reply be/src/exec/cross_join_node.cpp:56]: fix

File: be/src/exec/intersect_node.cpp [hunk 7]
```diff
@@ -44,6 +45,7 @@ Status IntersectNode::init(const TPlanNode& tnode, RuntimeState* state) {
 // repeat [2] this for all the rest child
 Status IntersectNode::open(RuntimeState* state) {
     RETURN_IF_ERROR(SetOperationNode::open(state));
+    SCOPED_SWITCH_THREAD_LOCAL_MEM_TRACKER_AND_ERR_CB(mem_tracker(), "Intersect , while probing the hash table.");
```
[the remark this answers is not in the document]
Comment (xinyiZzz) [reply be/src/exec/intersect_node.cpp:48]: fix

File: be/src/exec/except_node.cpp [hunk 8]
```diff
@@ -40,6 +41,7 @@ Status ExceptNode::init(const TPlanNode& tnode, RuntimeState* state) {
 
 Status ExceptNode::open(RuntimeState* state) {
     RETURN_IF_ERROR(SetOperationNode::open(state));
+    SCOPED_SWITCH_THREAD_LOCAL_MEM_TRACKER_AND_ERR_CB(mem_tracker(), "Except , while probing the hash table.");
```
[the remark this answers is not in the document]
Comment (xinyiZzz) [reply be/src/exec/except_node.cpp:44]: fix

File: be/src/runtime/thread_mem_tracker_mgr.h (hunk 2 above)
Comment (xinyiZzz) [reply be/src/runtime/thread_mem_tracker_mgr.h:75]: fix

File: be/src/olap/lru_cache.cpp [hunk 9]
```diff
@@ -491,11 +488,13 @@ Cache::Handle* ShardedLRUCache::lookup(const CacheKey& key) {
 }
 
 void ShardedLRUCache::release(Handle* handle) {
+    SCOPED_SWITCH_THREAD_LOCAL_MEM_TRACKER(_mem_tracker);
```
Comment (morningman) [inline be/src/olap/lru_cache.cpp]: Can we add a metric counter for this SWITCH operation?
I want to know how many SWITCH may be called.

File: be/src/vec/exec/vcross_join_node.cpp (hunk 4 above)
Comment (xinyiZzz) [reply be/src/vec/exec/vcross_join_node.cpp:57]: fix

File: be/src/vec/exec/vset_operation_node.cpp (hunk 5 above)
Comment (xinyiZzz) [reply be/src/vec/exec/vset_operation_node.cpp:233]: fix

File: be/src/runtime/thread_mem_tracker_mgr.cpp (hunk 6 above)
Comment (xinyiZzz) [reply be/src/runtime/thread_mem_tracker_mgr.cpp:75]: why is this, adding two strings directly seems to be more concise.

File: be/src/olap/lru_cache.cpp (hunk 9 above)
Comment (xinyiZzz) [reply be/src/olap/lru_cache.cpp]: ok, i try to add

Comment (xinyiZzz) [reply be/src/olap/lru_cache.cpp:492]: fix

File: be/src/olap/lru_cache.cpp [hunk 10]
```diff
@@ -449,11 +443,13 @@ ShardedLRUCache::ShardedLRUCache(const std::string& name, size_t total_capacity,
         : _name(name),
           _last_id(1),
           _mem_tracker(MemTracker::create_tracker(-1, name, nullptr, MemTrackerLevel::OVERVIEW)) {
+    SCOPED_SWITCH_THREAD_LOCAL_MEM_TRACKER(_mem_tracker);
     const size_t per_shard = (total_capacity + (kNumShards - 1)) / kNumShards;
     for (int s = 0; s < kNumShards; s++) {
         _shards[s] = new LRUCache(type);
         _shards[s]->set_capacity(per_shard);
     }
+    thread_local_ctx.get()->_thread_mem_tracker_mgr->clear_untracked_mems();
```
Comment (morningman) [inline be/src/olap/lru_cache.cpp:452]: Why clear untracked mem here?
you can add some comment.

Comment (xinyiZzz) [reply be/src/olap/lru_cache.cpp:454]: After the lru cache is created in the main thread, the main thread will not switch the lru cache mem tracker again.

After the non-query thread switches the mem tracker, if the thread will not switch the mem tracker again in the short term, can consider manually clear_untracked_mems.

The query thread will automatically clear_untracked_mems when detach_task.

**Review by morningman — APPROVED** [review]
LGTM

**Comment by morningman** [comment]
merge for quick test

**Pull request merged**